src/sort_filter/Hungarian.cpp

```diff
@@ -0,0 +1,454 @@
+///////////////////////////////////////////////////////////////////////////////
+// Hungarian.cpp: Implementation file for Class HungarianAlgorithm.
+//
+// This is a C++ wrapper with slight modification of a hungarian algorithm implementation by Markus Buehren.
+// The original implementation is a few mex-functions for use in MATLAB, found here:
+// http://www.mathworks.com/matlabcentral/fileexchange/6543-functions-for-the-rectangular-assignment-problem
+//
+// Both this code and the orignal code are published under the BSD license.
+// by Cong Ma, 2016
+//
+
+#include "Hungarian.h"
+
+using namespace std;
+
+HungarianAlgorithm::HungarianAlgorithm() {}
+HungarianAlgorithm::~HungarianAlgorithm() {}
+
+//********************************************************//
+// A single function wrapper for solving assignment problem.
+//********************************************************//
+double HungarianAlgorithm::Solve(
+	vector<vector<double>> &DistMatrix,
+	vector<int> &Assignment)
+{
+	unsigned int nRows = DistMatrix.size();
+	unsigned int nCols = DistMatrix[0].size();
+
+	double *distMatrixIn = new double[nRows * nCols];
+	int *assignment = new int[nRows];
+	double cost = 0.0;
+
+	// Fill in the distMatrixIn. Mind the index is "i + nRows * j".
+	// Here the cost matrix of size MxN is defined as a double precision array of N*M elements.
+	// In the solving functions matrices are seen to be saved MATLAB-internally in row-order.
+	// (i.e. the matrix [1 2; 3 4] will be stored as a vector [1 3 2 4], NOT [1 2 3 4]).
+	for (unsigned int i = 0; i < nRows; i++)
+		for (unsigned int j = 0; j < nCols; j++)
+			distMatrixIn[i + nRows * j] = DistMatrix[i][j];
+
+	// call solving function
+	assignmentoptimal(assignment, &cost, distMatrixIn, nRows, nCols);
+
+	Assignment.clear();
+	for (unsigned int r = 0; r < nRows; r++)
+		Assignment.push_back(assignment[r]);
+
+	delete[] distMatrixIn;
+	delete[] assignment;
+	return cost;
+}
+
+//********************************************************//
+// Solve optimal solution for assignment problem using Munkres algorithm, also known as Hungarian Algorithm.
+//********************************************************//
+void HungarianAlgorithm::assignmentoptimal(
+	int *assignment,
+	double *cost,
+	double *distMatrixIn,
+	int nOfRows,
+	int nOfColumns)
+{
+	double *distMatrix, *distMatrixTemp, *distMatrixEnd, *columnEnd, value, minValue;
+	bool *coveredColumns, *coveredRows, *starMatrix, *newStarMatrix, *primeMatrix;
+	int nOfElements, minDim, row, col;
+
+	/* initialization */
+	*cost = 0;
+	for (row = 0; row < nOfRows; row++)

```

Comment:
@BrennoCaldato Regarding Codacy's feedback on things like the scope of loop variables — I see it as a purely stylistic choice, ultimately, and in practice there will be no appreciable difference in performance between the two. However, I tend to side with the voices in [this discussion](http://www32.cplusplus.com/forum/general/116308/) who say that, _if anything_, multiple for loops that each use a separate, locally-scoped variable may run slightly _faster_ than reusing the same variable allocated once at the beginning.

That's mostly because, if we're talking about a POD variable (like a loop counter), the compiler can, and will, just store that counter in a register. Which means it won't require _any_ allocations at all. It's probably smart enough that it's going to do that **regardless** when optimization is turned on, but the version with locally-scoped variables gives it a better chance of optimizing away the allocations even at lower optimization levels.

See also Matt Godbolt's talk from CppCon 2017, where he shows off Compiler Explorer and uses it to examine how various styles of C++ loop code get translated into assembly instructions at different optimization levels: 

https://youtu.be/bSkpMdDe4g4

(The entire talk is really worth watching if you have time, it's all amazing. At around 35 minutes in, there's a segment where he demonstrates some of the tricks the compiler uses to optimize adds and multiplies, as well as to _avoid_ divide operations — that really blew me away.)

So, long story short, take these two bits of otherwise-identical code:

```c++
// Version 1
int row;
for (row = 0; row < ...; ++row) {
  ...
}
for (row = 0; ...) {
  ...
}
for (row = 0; ...) {
  ...
}
```
```c++
// Version 2
for (int row = 0; row < ...; ++row) {
  ...
}
for (int row = 0; ...) {
  ...
}
for (int row = 0; ...) {
  ...
}
```
In practice, all other things being equal:
* The difference in terms of real-world performance will be too small to measure, in basically all scenarios.
* With optimization turned on, both versions will most likely require _zero_ memory allocations for `row`.
* Therefore there's no practical advantage to Version 1 over the (IMHO) cleaner and more readable Version 2
* Codacy is therefore (again IMHO) correct to recommend rewriting instances of Version 1 as Version 2, where it can catch them.
* The same holds for instances that Codacy _can't_ catch. (It recommended reducing the scope of your `col`, but not `row`, even though `row` is never used outside of the loops either.)

But, as I said, ultimately I consider it a stylistic choice and I'm not hung up on imposing one or the other in the libopenshot code.

Reply:
@ferdnyc This is very interesting. I didn't have time to watch the video yet but I'll definitely do. Thanks for sending it.